### DISCUSSION TITLE
Error message on PC

### DISCUSSION THREAD

**Discussion by clave-sounds** [melanchall/drywetmidi#258] [General]
Hi Max. 

I've been running your API absolutely fine on my Mac. However a friend had problem on windows. I'm wondering if we can decipher the error. He sent it on Discord so couldn't send the whole error. Here's the bit he sent (I could ask him to email the whole error):

MidiDeviceException: Device is already in use (OUT_OPENRESULT_ALLOCATED).
Melanchall.DryWetMidi.Multimedia.NativeApi.HandleResult[TResult,TException] (TResult result, System.Func`3[T1,T2,TResult] createException) (at <1ee5422f8ce74a06831827bb5f773e0a>:0)
Melanchall.DryWetMidi.Multimedia.NativeApiUtilities.HandleDevicesNativeApiResult[TResult] (TResult result) (at <1ee5422f8ce74a06831827bb5f773e0a>:0)
Melanchall.DryWetMidi.Multimedia.OutputDevice.EnsureHandleIsCreated () (at <1ee5422f8ce74a06831827bb5f773e0a>:0)

He doesn't use MIDI, or even know what it is, so it's very unlikely he's using a MIDI output in another application. And he does have one (MIDI Port) because my app runs a scan for MIDI ports and came back with the classic Microsoft GM synth one.
So I'm guessing that the error is within unity. E.G the MIDI Port is in use by unity and being requested twice, but I don't get this problem on my Mac. I wondered if this could be something to do with disposing the OutputDevice after use in the scripts, I actually hadn't created any disposes because I'm using MIDI all the time in the app I'm making. Should I be disposing it after every MIDI event? Or after every scene? Or just on application close? Or do you think the problem is something else?

Thanks

**Comment by melanchall** [comment]
Hi,

> Should I be disposing it after every MIDI event? Or after every scene? Or just on application close?

Yes, you should dispose a MIDI device. And that's explicitly stated in the docs: https://melanchall.github.io/drywetmidi/articles/devices/Output-device.html

> You should always take care about disposing an OutputDevice, so use it inside `using` block or call `Dispose` manually. Without it all resources taken by the device will live until GC collect them via finalizer of the `OutputDevice`. It means that sometimes you will not be able to use different instances of the same device across multiple applications or different pieces of a program.

MIDI system on macOS is much better and you can use the same devices across multiple apps there. But Windows MIDI system provides only exclusive access to devices.

As for the moment when to dispose a device, you need to choose it by yourself. But that

> Should I be disposing it after every MIDI event?

is absolutely wrong. Just dispose a device when the app doesn't need it anymore. Common case is to dispose it when the app exits. In case of Unity - when a game has been stopped.

Thanks,
Max

**Reply by clave-sounds** [reply]
Ok thanks Max. Makes sense. I'll try that then.

**Comment by clave-sounds** [comment]
Hi Max. 

I think I have reached a dead-end due to my inexperience, but before I switch to another asset, I thought it might be worth asking your advice as you've been so helpful thus far. 

Firstly to explain my problem:
My app/game requires MIDI in one direction only: from the code generating MIDI to some sort of in-game sound font player and then unity AudioSource. I don't need any MIDI interaction with external devices or software. No Midi info needs to leave or enter the game.

In the short term I've been using DryWetMidi in unity and sending it via IAC bus to software on my Mac to playback GM Midi, but I knew that would have to be fixed at some point by bringing the sound generation inside the game, and that is where I've become stuck.

I realise this is not your responsibility as you have made a (fantastic) midi API which is responsible for Midi events only and not audio playback. However, I thought you would probably know if this is possible. If it is not possible, then no problem, I have learnt so much from doing this so it's not been wasted, I will have to go back many steps and refactor my code to work with another asset. However, maybe it is possible, but I'm just not sure.

So my question: Is it possible to link DryWetMidi with something like FluidSynth inside Unity, and if so, what is the mechanism I should research to do so? Would that be virtual ports? Or something else?

As always, thanks for your API, and your helpful replies :)

**Reply by melanchall** [reply]
Hi,

[Custom output device](https://melanchall.github.io/drywetmidi/articles/devices/Output-device.html#custom-output-device) is what you need I suppose. So you can use all the methods of DryWetMIDI just passing your custom "device" which will redirect messages to FluidSynth. Here you can see the same issue – https://github.com/melanchall/drywetmidi/issues/83.

**Reply by clave-sounds** [reply]
Thanks Max.

So if I understand correctly, it means using the event callback to separate the MIDI code and convert it into a format for the fluid synth and making your output device into a state so it doesn't try to link with any hardware/ software ports?

**Reply by melanchall** [reply]
You can use the callback, but custom output device is a more correct approach. You can view an example by the first link in my previous message. You device will look like that:

```csharp
private sealed class FluidSynthOutputDevice : IOutputDevice
{
    public event EventHandler<MidiEventSentEventArgs> EventSent;

    public void PrepareForEventsSending()
    {
    }

    public void SendEvent(MidiEvent midiEvent)
    {
        // Here you need to send midiEvent to FluidSynth
        // ...

        EventSent?.Invoke(this, new MidiEventSentEventArgs(midiEvent));
    }

    public void Dispose()
    {
    }
}
```

So yes, you need to _convert it into a format for the fluid synth_ but place that conversion in `FluidSynthOutputDevice`. Then you can use that "device" in all methods. For example, for playback:

```csharp
var playback = midiFile.GetPlayback(new FluidSynthOutputDevice());
```

**Comment by clave-sounds** [comment]
Nice!
That was actually very easy!
